src/api/forms/service.js

```diff
@@ -133,6 +134,46 @@ export async function updateDraftFormDefinition(formId, formDefinition) {
   return draftFormDefinition.create(formId, formDefinition)
 }
 
+/**
+ * Promotes a form from draft to live
+ * @param {string} formId - ID of the form
+ * @param {FormMetadataAuthor} author - the author of the promotion
+ */
+export async function promoteForm(formId, author) {
+  // Get the form metadata from the db
+  const form = await getForm(formId)
+
+  if (!form) {
+    throw Boom.notFound(`Form with id '${formId}' not found`)
+  }
+
+  // Build the live state
+  const now = new Date()
+  const state = {
+    updatedAt: now,
+    updatedBy: author
+  }
+
+  // Set the "created" state if this is the
+  // first time the form has been made live
+  if (!form.live) {
+    Object.assign(state, {
+      createdAt: now,
+      createdBy: author
+    })
+  }
```

Comment:
Because we're merging objects here you probably got type errors on `formMetadata.update()`

Mind if we create the "state" in one go?

Then it satisfies the `FormMetadataState` checks

```suggestion

  // Set the "created" state if this is the
  // first time the form has been made live
  const { createdAt, createdBy } = form.live ?? {
    createdAt: now,
    createdBy: author
  }

  /**
   * @satisfies {FormMetadataState}
   */
  const state = {
    createdAt,
    createdBy,
    updatedAt: now,
    updatedBy: author
  }
```

Reply:
~~No errors here :thinking: ~~

~~Seems happy with it~~

Ok, it is now you've spolied it with Partial<FormMetadataDocument>. :wink:

Reply:
Yep, caught out by my own `{object}` suggestion above 😆

But yep, this is the solution to `Partial<FormMetadataDocument>` woes

Using `{object}` works fine but we'd want compiler checks to know about potential patch restrictions, e.g. should we omit `'slug'` as an allowed key to be included in the patch param

Reply:
Unfortunately, these type gymastics :smile:  fundamentally change the intent of the code.
`createdAt` and `createdBy` are patched to the DB in your version whereas my intent is they shouldn't ever be updated.
I realise they're getting overwritten with the same value we just hauled out of the DB, but still they're no longer write-once then read-only.

Reply:
I'll have a look on Monday because I do like that `Partial<>`.

Reply:
Ahh that's fine too, we can make it work that way!

Would need to be `Partial< FormMetadata>` but the nested state was also `Partial<FormMetadataState>`

Yep let's make it work on Monday 🤯

Reply:
Resolved with 0ae458c0f776221ea89bb7b2948fa2300c092de5